### DISCUSSION TITLE
Use AST transforms to remove server code instead of tree shaking

### DISCUSSION THREAD

**Discussion by ryanflorence** [remix-run/remix#5374] [Proposals]
tl;dr

Stop relying on flaky tree shaking, use more robust AST parsing.

## Background

Today we remove server code from client bundles by creating "virtual modules" in front of the route module entry points to esbuild, basically something like this:

```jsx
export { default, meta } from "./routes/actual-module";
```

Then esbuild tree shakes out `loader` and `action`, etc.

## Problems

App developers need to follow a lot of rules, and so do their dependencies for this to work:

1. Can't import node built in modules like `fs` to use in loaders directly in route modules
2. Can't create HOF for loaders/actions/headers
3. Third party deps need to have `sideEffects: false` if imported into a route module
4. No module side effects

I could go on, but the main point is that practically every Remix app bumps into this problem almost immediately and there's no good place for Remix to jump in and help because the errors are thrown before Remix even exists.

For example, if you try to use `import fs from "fs/promises"` in a route module, your browser bundle ends up with a line like this:

```js
// app/routes/index.tsx
import fs from "fs/promises";
```

There is no way Remix can help with the error that will show up in the browser here. The same situation occurs with third party deps, so even if we marked all built-ins as "side-effect free", we're just playing whack-a-mole, there will always be new third party modules that will end up in client bundles that shouldn't be there.

## Solution

We can use AST transforms to remove the server code instead of tree shaking. This will be far more reliable and doesn't require app modules or third party modules to follow any special rules. The only way code will end up in client bundles is if they actually use the thing in the exports that go to the browser (the default component, meta, etc.).

We've already started parsing ASTs for HMR. As long as we're going to co-locate server and client code in the same module, we should do our best to make sure it gets shipped to the browser in the most reliable way. Tree shaking has not been as reliable as I had hoped it would be.

**Comment by pcattori** [comment]
I think we'll still need to use treeshaking (and marking dependencies as `sideEffects: false`) to remove unused (i.e. server-only) imports for the browser build. E.g.

```ts
import fs from 'node:fs/promises'

export let loader = async () => {
  let data = await fs.readFile('blah.txt', 'utf8')
  return json({ data })
}

export default () => {
  let { data } = useLoaderData<typeof loader>()
  return <pre>{data}</pre>
}
```

Even if the AST removes `export let loader = {...}`, we still need to rely on treeshaking to remove the `fs` import.
Same would be true for a `createLoader` paradigm; still need to treeshake out the top-level `fs` import.

**Reply by ryanflorence** [reply]
We'd track imported modules and if they aren't used in the client exports, remove them.

**Reply by pcattori** [reply]
That would work! Still have to (implicitly) treat imports as side-effect free, but there's no way to get around that anyway when separating out the browser build. 👍

**Reply by pcattori** [reply]
There's additional complexity though if there's indirect usage of an import:

```ts
import fs from 'node:fs/promises'

let helper1 = async (filename: string) => fs.readFile(filename, 'utf8')

let helper2 = async (filename: string) => {
  if (filename.startsWith('foobar')) throw Error('no foobar plz')
  return helper1(filename)
}

export let loader = async () => {
  let data = await helper2('blah.txt')
  return json({ data })
}

export default () => {/* same as before */}
```

In this case we'd need to reimplement treeshaking/dead code elimination if we want to reliably remove `fs` import with AST transforms.

So we'd have to traverse references, build a reference tree, and then prune it. That would let us do the treeshaking/dead code elimination reliably in AST transforms, but that's additional complexity we'd be taking on.

**Comment by mjackson** [comment]
Just to be clear here, by "parse the AST" what you really mean is "create our own bespoke treeshaking algorithm".

I have a few concerns with this idea:

- Yes, our own treeshaking algorithm would give us more control over the process of eliminating code, but esbuild is already a very capable dead code eliminator. I don't think we are using it as effectively as we could be. For example, we are not currently using esbuild's ability to mark a module as `sideEffects: false` [in a plugin](https://esbuild.github.io/plugins/#on-resolve-results), which I believe could help us mark all 3rd party deps as side-effect free if we wanted to.
- By creating our own treeshaking algorithm we could potentially introduce JavaScript semantics that are unexpected. So we get further away from being "just JS" and venture into "custom language" territory. For example, automatically removing calls to higher-order functions with side effects, although desirable for our purposes, would be an unexpected change to the JS language semantics.

I'm not totally opposed to doing our own treeshaking, but I would like to examine the problems with our current implementation on top of esbuild and see if we can address them within the constraints of our current architecture before we ditch it and do it ourselves.

**Reply by pcattori** [reply]
Right, there are two concerns here:
1. Surgically removing Remix-specific code like `loader` and `action`
2. Treeshaking / DCE _after_ (1) is done

I think tackling (1) with AST transforms has some definite upside over using virtual modules. I'm not convinced about doing (2) ourselves in AST transforms if `esbuild` can do that for us.

**Reply by mjackson** [reply]
> Surgically removing Remix-specific code like loader and action

Is the goal to enable higher-order functions to be used to create loader and action functions? I think I could get behind that idea. Everyone already knows those don't make it to the browser anyway, so they shouldn't be surprised when any side-effects in their `createLoader()` function don't run in the browser.

BTW I think we shouldn't need to track `import`s in this case. esbuild should be very able to remove any unused imports. I don't want to overlap too much with it.

**Reply by pcattori** [reply]
> Is the goal to enable higher-order functions to be used to create loader and action functions?

That's where my head is at, yes. Though not sure if @ryanflorence has other primary reasons for wanting this.

**Comment by ricsam** [comment]
I'll just spitball a little here:

I think the main driving point for this is the HOF. Those allow the app to scale. When you create a route with many routes the module API constrains becomes really annoying, and the limitations become a blocker for further development.

### 1
In the routes could one not define `${filename}.server.ts` which exclusively exports loader and action? Within that file one could say:

```tsx
// userId$/invoices/invoiceId$.server.ts
export const { loader, action } = invoiceRoute({ where: { field: 'userId', value: 'userId$' }});
```

```tsx
// invoices/invoiceId$.server.ts
export const { loader, action } = invoiceRoute();
```

This doesn't work today, but it feels like it should?

### 2
Or even better would be if I could define all of the react-router / module API exports with just one line:


```tsx
// userId$/invoices/invoiceId$.ts
export const { loader, action, handle, component, links } = invoiceRoute({ where: { field: 'userId', value: 'userId$' }});
```

```tsx
// invoices/invoiceId$.ts
export const { loader, action, handle, component, links } = invoiceRoute();
```

Ok. For this AST transforms would be nice.

### 3
I might have a lot of different invoice list pages, by: country, company, department, user.

With that it would be ideal if one could define a file with routes:

```tsx
// app/routes.tsx
export const routes = {
  'invoices': invoices(),
  'userId$/invoices': invoices({ where: { field: 'userId', value: 'userId$' }}),
  'countryId$/invoices': invoices({ where: { field: 'country', value: 'countryId$' }})
  // ...
}
```
Like a createBrowserRouter in react-router v6.

Each route probably has different layout, e.g. `/invoices` is just a table and the `/userId$/invoices` is the user profile page, with the invoices as a child route.

I've sort of hit the ceiling with remix in the app I am building now where I really feel like I would need hofs, because the routing becomes pretty complex.

In general in big projects the routing can become a beast on its own, at some point one needs abstractions (on abstractions) to handle the routing, and it would be great if that was supported out of the box.

**Comment by brophdawg11** [comment]
Adding this for visibility when we get to this work.  This seems to cause issues with our current approach so we should ensure the AST approach handles it.  When the `loader` is exported separately at the bottom of the file, our `json` import is not properly removed, causing a `Dynamic require of "fs/promises" is not supported` error when we try to import the client bundle

```jsx
import { json } from "@remix-run/node"

const loader = async () => json({posts: []});

const Route = () => <p>route</p>;

export default Route
export {loader}
```
